packages/blocks/src/api/parser/index.js

```diff
@@ -0,0 +1,267 @@
+ *
+ * @param {WPRawBlock} rawBlock
+ *
+ * @return {WPRawBlock} The block's name and attributes, changed accordingly if a match was found
+ */
+function convertLegacyBlocks( rawBlock ) {
+	const [
+		correctName,
+		correctedAttributes,
+	] = convertLegacyBlockNameAndAttributes(
+		rawBlock.blockName,
+		rawBlock.attrs
+	);
+	return {
+		...rawBlock,
+		blockName: correctName,
+		attrs: correctedAttributes,
+	};
+}
+
+/**
+ * Normalize the raw block by applying the fallback block name if none given,
+ * sanitize the parsed HTML...
+ *
+ * @param {WPRawBlock} rawBlock The raw block object.
+ *
+ * @return {WPRawBlock} The normalized block object.
+ */
+export function normalizeRawBlock( rawBlock ) {
+	const fallbackBlockName = getFreeformContentHandlerName();
+
+	// If the grammar parsing don't produce any block name, use the freeform block.
+	const rawBlockName = rawBlock.blockName || getFreeformContentHandlerName();
+	const rawAttributes = rawBlock.attrs || {};
+	const rawInnerBlocks = rawBlock.innerBlocks || [];
+	let rawInnerHTML = rawBlock.innerHTML.trim();
+
+	// Fallback content may be upgraded from classic content expecting implicit
+	// automatic paragraphs, so preserve them. Assumes wpautop is idempotent,
+	// meaning there are no negative consequences to repeated autop calls.
+	if ( rawBlockName === fallbackBlockName ) {
+		rawInnerHTML = autop( rawInnerHTML ).trim();
+	}
+
+	return {
+		...rawBlock,
+		blockName: rawBlockName,
+		attrs: rawAttributes,
+		innerHTML: rawInnerHTML,
+		innerBlocks: rawInnerBlocks,
+	};
+}
+
+/**
+ * Uses the "unregistered blockType" to create a block object.
+ *
+ * @param {WPRawBlock} rawBlock block.
+ *
+ * @return {WPRawBlock} The unregistered block object.
+ */
+function createMissingBlockType( rawBlock ) {
+	const unregisteredFallbackBlock =
+		getUnregisteredTypeHandlerName() || getFreeformContentHandlerName();
+
+	// Preserve undelimited content for use by the unregistered type
+	// handler. A block node's `innerHTML` isn't enough, as that field only
+	// carries the block's own HTML and not its nested blocks.
+	const originalUndelimitedContent = serializeRawBlock( rawBlock, {
+		isCommentDelimited: false,
+	} );
+
+	// Preserve full block content for use by the unregistered type
+	// handler, block boundaries included.
+	const originalContent = serializeRawBlock( rawBlock, {
+		isCommentDelimited: true,
+	} );
+
+	return {
+		blockName: unregisteredFallbackBlock,
+		attrs: {
+			originalName: rawBlock.blockName,
+			originalContent,
+			originalUndelimitedContent,
+		},
+		innerHTML: rawBlock.blockName ? originalContent : rawBlock.innerHTML,
+		innerBlocks: rawBlock.innerBlocks,
+		innerContent: rawBlock.innerContent,
+	};
+}
+
+/**
+ * Given a raw block returned by grammar parsing, returns a fully parsed block.
+ *
+ * @param {WPRawBlock} rawBlock The raw block object.
+ *
+ * @return {WPBlock} Fully parsed block.
+ */
+export function parseRawBlock( rawBlock ) {
```

Comment:
In the last commit I did an extensive refactoring of the parsing code. The idea is that now you can take a look at the `parseRawBlock` function which used to be named `createBlockWithFallback` and understand the different steps of the parsing operation. 

I also added the `WPBlock` and the `WPRawBlock` types to understand the inputs and outputs of each step properly.

Validation and fixing is still a bit complex, but not sure we can do a lot better there for the moment.

Reply:
Obviously a "clear code" is subjective and I think we might need a detailed schema to properly understand how parsing works but I'd love feedback from folks about how to make the different steps clearer.